File: deck_game_installer/steam.py
import os
import struct
import vdf
from pathlib import Path
from typing import Optional, Dict, Any, List

class SteamManager:
    def __init__(self):
        self.steam_path = Path.home() / ".local/share/Steam"
        self.userdata_path = self.steam_path / "userdata"
        self.user_ids = self._find_user_ids()
        self.user_id = self.user_ids[0] if self.user_ids else "0"

    def _find_user_ids(self) -> list:
        if not self.userdata_path.exists():
            return []
        # Return all numeric subdirs
        return [d.name for d in self.userdata_path.iterdir() if d.is_dir() and d.name.isdigit()]

    def get_shortcuts_path(self) -> Path:
        return self.userdata_path / self.user_id / "config/shortcuts.vdf"

    def get_config_path(self) -> Path:
        """Returns the path to the global config.vdf where compatibility settings are stored."""
        return self.steam_path / "config/config.vdf"

    @staticmethod
    def generate_app_id(exe_path: str, app_name: str) -> int:
        """Generates the 32-bit AppID used by Steam for non-Steam games."""
        key = f"{exe_path}{app_name}"
        from zlib import crc32
        u32 = (crc32(key.encode('utf-8')) | 0x80000000) & 0xFFFFFFFF
        return struct.unpack('<i', struct.pack('<I', u32))[0]
    
    @staticmethod
    def get_url_app_id_from_u32(u32: int) -> str:
        """Generates the 64-bit AppID string for steam:// URLs from a 32-bit AppID."""
        # Ensure it's treated as unsigned for the shift. 
        # Steam stores it as signed 32-bit in binary VDF.
        u32_unsigned = struct.unpack('<I', struct.pack('<i', u32))[0] if u32 < 0 else u32
        u64 = (u32_unsigned << 32) | 0x02000000
        return str(u64)

    def find_app_id_by_path(self, exe_path: str) -> Optional[int]:
        """Finds the 32-bit AppID for an existing shortcut by searching all userdata directories."""
        target_exe = f'"{exe_path}"'
        # Also try without quotes just in case
        target_exe_alt = exe_path
        
        for uid in self.user_ids:
            sp = self.userdata_path / uid / "config/shortcuts.vdf"
            if not sp.exists(): continue
            try:
                with open(sp, "rb") as f:
                    shortcuts = vdf.binary_load(f)
                items = shortcuts.get("shortcuts", {})
                for item in items.values():
                    if item.get("Exe") == target_exe or item.get("Exe") == target_exe_alt:
                        return item.get("appid")
            except Exception:
                continue
        return None

    def add_shortcut(self, app_name: str, exe_path: str, arguments: str = "", start_dir: str = "") -> int:
        """Adds a non-Steam game shortcut directly to shortcuts.vdf."""
        sp = self.get_shortcuts_path()
        shortcuts = {"shortcuts": {}}
        
        if sp.exists():
            try:
                with open(sp, "rb") as f:
                    shortcuts = vdf.binary_load(f)
            except Exception:
                pass

        if "shortcuts" not in shortcuts:
            shortcuts["shortcuts"] = {}

        # Check if already exists
        target_exe = f'"{exe_path}"'
        for entry in shortcuts["shortcuts"].values():
            if entry.get("Exe") == target_exe or entry.get("Exe") == exe_path:
                return entry.get("appid")

        # Create new entry
        appid = self.generate_app_id(exe_path, app_name)
        new_index = str(len(shortcuts["shortcuts"]))
        
        shortcuts["shortcuts"][new_index] = {
            "appid": appid,
            "AppName": app_name,
            "Exe": target_exe,
            "StartDir": start_dir or str(Path(exe_path).parent),
            "icon": "",
            "ShortcutPath": "",
            "LaunchOptions": arguments,
            "IsHidden": 0,
            "AllowDesktopConfig": 1,
            "AllowOverlay": 1,
            "OpenVR": 0,
            "Devkit": 0,
            "DevkitGameID": "",
            "DevkitOverrideAppID": 0,
            "LastPlayTime": 0,
            "FlatpakAppID": "",
            "sortas": "",
            "tags": {}
        }

        # Backup and Save
        if sp.exists():
            import shutil
            shutil.copy2(sp, sp.with_suffix(".vdf.bak"))

        with open(sp, "wb") as f:
            vdf.binary_dump(shortcuts, f)

        return appid

    def get_proton_version(self, app_id_32: int) -> Optional[str]:
        """Gets the currently configured Proton version for a specific AppID, if any."""
        path = self.get_config_path()
        if not path.exists():
            return None

        try:
            with open(path, "r") as f:
                config = vdf.load(f)
            
            mapping = config.get('InstallConfigStore', {}).get('Software', {}).get('Valve', {}).get('Steam', {}).get('CompatToolMapping', {})
            
            # Convert signed to unsigned AppID
            u32_unsigned = struct.unpack('<I', struct.pack('<i', app_id_32))[0] if app_id_32 < 0 else app_id_32
            
            entry = mapping.get(str(u32_unsigned))
            if entry:
                return entry.get('name')
        except Exception:
            pass
        return None

    def set_proton_version(self, app_id_32: int, proton_version: str):
        """Sets the Proton version for a specific AppID in config.vdf."""
        path = self.get_config_path()
        if not path.exists():
            return

        with open(path, "r") as f:
            config = vdf.load(f)

        # Path: InstallConfigStore/Software/Valve/Steam/CompatToolMapping
        try:
            mapping = config['InstallConfigStore']['Software']['Valve']['Steam']['CompatToolMapping']
        except KeyError:
            # Create the path if it doesn't exist
            if 'InstallConfigStore' not in config: config['InstallConfigStore'] = {}
            if 'Software' not in config['InstallConfigStore']: config['InstallConfigStore']['Software'] = {}
            if 'Valve' not in config['InstallConfigStore']['Software']: config['InstallConfigStore']['Software']['Valve'] = {}
            if 'Steam' not in config['InstallConfigStore']['Software']['Valve']: config['InstallConfigStore']['Software']['Valve']['Steam'] = {}
            if 'CompatToolMapping' not in config['InstallConfigStore']['Software']['Valve']['Steam']: 
                config['InstallConfigStore']['Software']['Valve']['Steam']['CompatToolMapping'] = {}
            mapping = config['InstallConfigStore']['Software']['Valve']['Steam']['CompatToolMapping']

        # Steam expects unsigned 32-bit AppID strings in CompatToolMapping
        u32_unsigned = struct.unpack('<I', struct.pack('<i', app_id_32))[0] if app_id_32 < 0 else app_id_32
        
        mapping[str(u32_unsigned)] = {
            'name': proton_version,
            'config': '',
            'priority': '250'
        }

        with open(path, "w") as f:
            vdf.dump(config, f)

    def restart_steam(self):
        """Attempts to restart Steam to apply config changes."""
        import subprocess
        # Try to kill Steam
        subprocess.run(["pkill", "-x", "steam"], capture_output=True)
        # Give it a moment to die
        import time
        time.sleep(3)
        # Relaunch Steam
        subprocess.Popen(["steam"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

File: steam/steam.go
package steam

import (
	"errors"
	"hash/crc32"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"syscall"
	"time"

	"deck-game-installer/vdf"
)

type Manager struct {
	steamPath    string
	userdataPath string
	userIDs      []string
	userID       string
}

func NewManager() *Manager {
	home, _ := os.UserHomeDir()
	steamPath := filepath.Join(home, ".local", "share", "Steam")
	userdata := filepath.Join(steamPath, "userdata")

	userIDs := findUserIDs(userdata)
	userID := "0"
	if len(userIDs) > 0 {
		userID = userIDs[0]
	}

	return &Manager{
		steamPath:    steamPath,
		userdataPath: userdata,
		userIDs:      userIDs,
		userID:       userID,
	}
}

func findUserIDs(userdata string) []string {
	entries, err := os.ReadDir(userdata)
	if err != nil {
		return nil
	}
	var ids []string
	for _, e := range entries {
		if e.IsDir() && isDigits(e.Name()) {
			ids = append(ids, e.Name())
		}
	}
	return ids
}

func isDigits(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return s != ""
}

func (m *Manager) ShortcutsPath() string {
	return filepath.Join(m.userdataPath, m.userID, "config", "shortcuts.vdf")
}

func (m *Manager) ConfigPath() string {
	return filepath.Join(m.steamPath, "config", "config.vdf")
}

func GenerateAppID(exePath, appName string) int32 {
	key := exePath + appName
	u32 := crc32.ChecksumIEEE([]byte(key)) | 0x80000000
	return int32(u32)
}

func GetURLAppIDFromU32(u32 int32) string {
	unsigned := uint32(u32)
	u64 := (uint64(unsigned) << 32) | 0x02000000
	return strconv.FormatUint(u64, 10)
}

func (m *Manager) FindAppIDByPath(exePath string) (int32, error) {
	quoted := "\"" + exePath + "\""
	for _, uid := range m.userIDs {
		path := filepath.Join(m.userdataPath, uid, "config", "shortcuts.vdf")
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		obj, err := ReadBinaryVDF(data)
		if err != nil {
			continue
		}
		shortcuts, ok := obj["shortcuts"].(map[string]kvValue)
		if !ok {
			continue
		}
		for _, v := range shortcuts {
			entry, ok := v.(map[string]kvValue)
			if !ok {
				continue
			}
			exe, _ := entry["Exe"].(string)
			if exe == quoted || exe == exePath {
				if appid, ok := entry["appid"].(int32); ok {
					return appid, nil
				}
			}
		}
	}
	return 0, errors.New("not found")
}

func (m *Manager) AddShortcut(appName, exePath, args, startDir string) (int32, error) {
	path := m.ShortcutsPath()

	obj := map[string]kvValue{"shortcuts": map[string]kvValue{}}
	if data, err := os.ReadFile(path); err == nil {
		if parsed, err := ReadBinaryVDF(data); err == nil {
			obj = parsed
		}
	}

	shortcuts, ok := obj["shortcuts"].(map[string]kvValue)
	if !ok {
		shortcuts = map[string]kvValue{}
		obj["shortcuts"] = shortcuts
	}

	quoted := "\"" + exePath + "\""
	var existingKey string

	for k, v := range shortcuts {
		entry, ok := v.(map[string]kvValue)
		if !ok {
			continue
		}
		exe, _ := entry["Exe"].(string)
		if exe == quoted || exe == exePath {
			if appid, ok := entry["appid"].(int32); ok {
				return appid, nil
			}
		}
		
		// If name matches, we'll maintain this entry instead of creating a new one
		name, _ := entry["AppName"].(string)
		if name == appName {
			existingKey = k
		}
	}

	appid := GenerateAppID(exePath, appName)
	var key string
	
	if existingKey != "" {
		key = existingKey
	} else {
		idx := nextShortcutIndex(shortcuts)
		key = strconv.Itoa(idx)
	}

	if startDir == "" {
		startDir = filepath.Dir(exePath)
	}

	shortcuts[key] = map[string]kvValue{
		"appid":             appid,
		"AppName":           appName,
		"Exe":               quoted,
		"StartDir":          startDir,
		"icon":              "",
		"ShortcutPath":      "",
		"LaunchOptions":     args,
		"IsHidden":          int32(0),
		"AllowDesktopConfig": int32(1),
		"AllowOverlay":      int32(1),
		"OpenVR":            int32(0),
		"Devkit":            int32(0),
		"DevkitGameID":      "",
		"DevkitOverrideAppID": int32(0),
		"LastPlayTime":      int32(0),
		"FlatpakAppID":      "",
		"sortas":            "",
		"tags":              map[string]kvValue{},
	}

	_ = os.MkdirAll(filepath.Dir(path), 0o755)
	data, err := WriteBinaryVDF(obj)
	if err != nil {
		return 0, err
	}
	if err := os.WriteFile(path, data, 0o644); err != nil {
		return 0, err
	}

	return appid, nil
}

func (m *Manager) UpdateShortcut(appID int32, newExePath, newStartDir string) error {
	path := m.ShortcutsPath()

	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	obj, err := ReadBinaryVDF(data)
	if err != nil {
		return err
	}

	shortcuts, ok := obj["shortcuts"].(map[string]kvValue)
	if !ok {
		return errors.New("shortcuts not found")
	}

	// Find the shortcut with matching appID
	for _, v := range shortcuts {
		entry, ok := v.(map[string]kvValue)
		if !ok {
			continue
		}
		if existingID, ok := entry["appid"].(int32); ok && existingID == appID {
			// Update the exe and start directory
			entry["Exe"] = "\"" + newExePath + "\""
			if newStartDir == "" {
				newStartDir = filepath.Dir(newExePath)
			}
			entry["StartDir"] = newStartDir
			
			// Write back to file
			newData, err := WriteBinaryVDF(obj)
			if err != nil {
				return err
			}
			return os.WriteFile(path, newData, 0o644)
		}
	}

	return errors.New("shortcut not found")
}

func nextShortcutIndex(shortcuts map[string]kvValue) int {
	max := -1
	for k := range shortcuts {
		if n, err := strconv.Atoi(k); err == nil && n > max {
			max = n
		}
	}
	return max + 1
}

func (m *Manager) GetProtonVersion(appID int32) (string, error) {
	path := m.ConfigPath()
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}

	root, err := vdf.Parse(string(data))
	if err != nil {
		return "", err
	}

	mapping := nestedMap(root, "InstallConfigStore", "Software", "Valve", "Steam", "CompatToolMapping")
	if mapping == nil {
		return "", errors.New("mapping not found")
	}

	unsigned := uint32(appID)
	key := strconv.FormatUint(uint64(unsigned), 10)
	entry, ok := mapping[key].(map[string]any)
	if !ok {
		return "", errors.New("entry not found")
	}
	name, _ := entry["name"].(string)
	if name == "" {
		return "", errors.New("name not found")
	}
	return name, nil
}

func (m *Manager) SetProtonVersion(appID int32, protonVersion string) error {
	path := m.ConfigPath()
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	root, err := vdf.Parse(string(data))
	if err != nil {
		return err
	}

	mapping := ensureNestedMap(root, "InstallConfigStore", "Software", "Valve", "Steam", "CompatToolMapping")
	unsigned := uint32(appID)
	key := strconv.FormatUint(uint64(unsigned), 10)

	mapping[key] = map[string]any{
		"name":     protonVersion,
		"config":   "",
		"priority": "250",
	}

	out := vdf.Dump(root)
	return os.WriteFile(path, []byte(out), 0o644)
}

func nestedMap(root map[string]any, keys ...string) map[string]any {
	current := root
	for _, k := range keys {
		v, ok := current[k]
		if !ok {
			return nil
		}
		next, ok := v.(map[string]any)
		if !ok {
			return nil
		}
		current = next
	}
	return current
}

func ensureNestedMap(root map[string]any, keys ...string) map[string]any {
	current := root
	for _, k := range keys {
		v, ok := current[k]
		if !ok {
			child := map[string]any{}
			current[k] = child
			current = child
			continue
		}
		next, ok := v.(map[string]any)
		if !ok {
			child := map[string]any{}
			current[k] = child
			current = child
			continue
		}
		current = next
	}
	return current
}

func (m *Manager) RestartSteam() {
	// Kill Steam gracefully first
	_ = exec.Command("steam", "-shutdown").Run()
	time.Sleep(2 * time.Second)
	
	// Force kill if still running
	_ = exec.Command("pkill", "-x", "steam").Run()
	time.Sleep(1 * time.Second)
	
	// Start Steam in background, fully detached from this process
	cmd := exec.Command("steam")
	cmd.Stdin = nil
	cmd.Stdout = nil
	cmd.Stderr = nil
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setsid: true, // Create new session to detach from parent
	}
	_ = cmd.Start()
	
	// Release the process so it's not tied to this parent
	if cmd.Process != nil {
		_ = cmd.Process.Release()
	}
}

func execCommand(name string, args ...string) error {
	cmd := exec.Command(name, args...)
	return cmd.Start()
}
